assets/app/features/upgrade/upgrade.controller.spec.js

```diff
@@ -44,14 +54,51 @@ describe('Upgrade Controller', function () {
             should.exist(controller.steps);
         });
 
-        it('should contains a nextStep function ', function() {
-            should.exist(controller.steps.nextStep);
-            expect(controller.steps.nextStep).toEqual(jasmine.any(Function));
+        describe('nextStep function', function () {
+            it('should exist', function () {
+                should.exist(controller.steps.nextStep);
+                expect(controller.steps.nextStep).toEqual(jasmine.any(Function));
+            });
+
+            it('when called it should move to the next step', function () {
+                expect($state.current.id).toBe(0);
+                // calculate the next step in the list
+                var nextStep = (controller.steps.list[upgradeStepsFactory.activeStep.id + 1]).state;
+                controller.steps.nextStep();
+                expect($state.go).toHaveBeenCalledWith(nextStep);
+            });
+
+            it('when called in the last step it should not change steps', function () {
+                // set the last step in the list as current
+                $state.current = {id: 0, name: 'upgrade.upgrade-nodes', state: 'upgrade.upgrade-nodes'};
+                // refresh the list so it sets it as active
+                upgradeStepsFactory.refeshStepsList();
```

Comment:
But mocking that part means that if it ever changes, the tests will continue to run for `nextStep` and `isLastStep` and they will return as valid. Both functions require that `upgradeStepsFactory.refeshStepsList()` works as intended to work properly so not sure why we will mock a dependent service instead of relaying on the real one to be sure that it works?

Mocking is the devil, it can hide real issues if you mock too much so I try to mock only the necessary stuff, hence the change to ```spyOn(upgradeStepsFactory, 'refeshStepsList').and.callThrough();``` as we need the *real* service to work for the other functions to do so.